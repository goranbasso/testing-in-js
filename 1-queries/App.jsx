import React from "react"
import PropTypes from "prop-types"

const App = (props) => {
  return (
    <div>
      Hello {props.name} <br />
      Welcome to {props.place}
    </div>)
}

App.propTypes = {
  name: PropTypes.string,
  place: PropTypes.string
}

App.defaultProps = {
  name: "World",
  place: 'other'
}

export default App
