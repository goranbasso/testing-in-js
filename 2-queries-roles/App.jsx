import React from "react"
import PropTypes from "prop-types"

const App = (props) => {
  return (
    <div>
      <p>Hello {props.name}!</p>
      <p>Please sign in.</p>
      <form>
        <label htmlFor="username">Username: </label>
        <input id="username" placeholder="username"/>
        <label htmlFor="password">Password: </label>
        <input id="password" placeholder="password" type="password"/>
        <button>{props.name}</button> <br />
        <a href={props.link}>Cool link</a>
        <a href={props.otherLink}>Other cool link</a>
        <a href={props.anotherLink}>Another cool link</a>
      </form>
    </div>
  )
}

App.propTypes = {
  name: PropTypes.string,
  link: PropTypes.string,
  otherLink: PropTypes.string,
  anotherLink: PropTypes.string
}

App.defaultProps = {
  name: "Sign in",
  link: "https://www.google.com",
  otherLink: "https://www.nrk.no",
  anotherLink: "https://developer.mozilla.org"
}

export default App
