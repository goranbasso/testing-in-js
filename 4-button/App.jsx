import React, {useEffect, useState} from 'react'

export const ghost = 'ðŸ‘»'
export const cat = 'ðŸ˜¸'

const App = () => {

  const [spooky, setSpooky] = useState([])

  useEffect(() => {
    setSpooky(false)
  }, [])

  const toggleSpooky = () => {
    setSpooky(!spooky)
  }

  const setSpookyExplicit = (state) => {
    setSpooky(state)
  }

  const buttonDisplayText = () => {
    return spooky ? ghost : cat
  }

  return (
    <div>
      <button onClick={toggleSpooky} type={"button"} data-testid={"spooky-button"}>{buttonDisplayText()}</button>
      <br />
      <button onClick={() => setSpookyExplicit(false)} type={"button"} data-testid={"spooky-false-button"}>{cat}</button>
      <button onClick={() => setSpookyExplicit(true)} type={"button"} data-testid={"spooky-true-button"}>{ghost}</button>
    </div>
  )
}

export default App