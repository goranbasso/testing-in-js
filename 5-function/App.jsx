import React, {useEffect, useState} from 'react'
import PropTypes from 'prop-types'
import {correctLoginFunction, redirectFunction, wrongLoginFunction} from './login'

const App = (props) => {

  const [username, setUsername] = useState([])
  const [password, setPassword] = useState([])
  const [loginSuccess, setLoginSuccess] = useState(false)

  useEffect(() => {
    setUsername('')
    setPassword('')
  }, [])

  const onUsernameChanged = (event) => {
    setUsername(event.target.value)
  }

  const onPasswordChanged = (event) => {
    const obfuscatedPassword = event.target.value.replace(/./g, '*')
    setPassword(obfuscatedPassword)
  }

  const onLoginClicked = () => {
    setLoginSuccess(props.login(username, password))
    if (props.login(username, password)) {
      props.redirect()
    }
  }

  return (
    <div>
      <label>
        Brukernavn:
        <input value={username} onChange={onUsernameChanged} data-testid={"username-input"} />
      </label>
      <br />
      <label>
        Passord:
        <input value={password} onChange={onPasswordChanged} data-testid={"password-input"} />
      </label>
      <br />
      <button type={"button"} onClick={onLoginClicked}>Login</button>
    </div>
  )
}

App.propTypes = {
  login: PropTypes.func.isRequired,
  redirect: PropTypes.func.isRequired
}

App.defaultProps = {
  login: correctLoginFunction,
  redirect: redirectFunction
}

export default App