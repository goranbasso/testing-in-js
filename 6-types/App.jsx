import React, {useEffect, useState} from 'react'
import PropTypes from 'prop-types'
import {difference, division, exponentiation, getDate, sum} from './helpers'
import {product} from "../0-warmup/functions"

const App = (props) => {

  const [paramA, setParamA] = useState([])
  const [paramB, setParamB] = useState([])
  const [result, setResult] = useState({
    sum: '',
    difference: '',
    product: '',
    division: '',
    exponentiation: ''
  })
  const [date, setDate] = useState([])

  useEffect(() => {
    setParamA('0')
    setParamB('0')
  }, [])

  const onParamAChanged = (event) => {
    setParamA(event.target.value)
  }

  const onParamBChanged = (event) => {
    setParamB(event.target.value)
  }

  const onCalculate = () => {
    const numberParamA = parseInt(paramA)
    const numberParamB = parseInt(paramB)
    const calculated = {
      sum: props.calculateSum(numberParamA, numberParamB),
      difference: props.calculateDifference(numberParamA, numberParamB),
      product: props.calculateProduct(numberParamA, numberParamB),
      division: props.calculateDivision(numberParamA, numberParamB),
      exponentiation: props.calculateExponentiation(numberParamA, numberParamB)
    }
    setResult(calculated)
  }

  const onGetDate = () => {
    setDate(props.getDate().toString())
  }

  return (
    <div>
      <label>
        Param A:
        <input value={paramA} onChange={onParamAChanged} data-testid={"param-a-input"} />
      </label>
      <br />
      <label>
        Param B:
        <input value={paramB} onChange={onParamBChanged} data-testid={"param-b-input"} />
      </label>
      <br />
      <button type={"button"} onClick={onCalculate}>calculate!</button>
      <br />

      <label htmlFor={"sum-result"}>
        Sum result:
      </label>
      <input id={"sum-result"} disabled value={result.sum} />
      <br />
      <label htmlFor={"diff-result"}>
        Difference result:
      </label>
      <input id={"diff-result"} disabled value={result.difference} />
      <br />
      <label htmlFor={"product-result"}>
        Product result:
      </label>
      <input id={"product-result"} disabled value={result.product} />
      <br />
      <label htmlFor={"divison-result"}>
        Division result:
      </label>
      <input id={"division-result"} disabled value={result.division} />
      <br />
      <label htmlFor={"power-result"}>
        Exponentiation result:
      </label>
      <input id={"power-result"} disabled value={result.exponentiation} />
      <br />

      <button type={"button"} onClick={onGetDate}>Get the date</button>
      <div>{date}</div>
    </div>
  )
}

App.propTypes = {
  calculateSum: PropTypes.func,
  calculateDifference: PropTypes.func,
  calculateProduct: PropTypes.func,
  calculateDivision: PropTypes.func,
  calculateExponentiation: PropTypes.func,
  getDate: PropTypes.func
}

App.defaultProps = {
  calculateSum: sum,
  calculateDifference: difference,
  calculateProduct: product,
  calculateDivision: division,
  calculateExponentiation: exponentiation,
  getDate: getDate
}

export default App