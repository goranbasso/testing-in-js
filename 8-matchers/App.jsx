import React, {useEffect, useState} from 'react';
import PropTypes from "prop-types";

export const getNumbersFromZeroToN = (n) => {
  return Array.from(Array(n + 1).keys())
}

const App = (props) => {

  const [rows, setRows] = useState([])

  useEffect(() => {
    const createRows = () => {
      setRows(getNumbersFromZeroToN(props.rowCount))
    }
    createRows()
  }, [])

  const product = (a, b) => a + b

  const renderRows = () => {
    const rowElements = []
    rows.forEach((row, index) => {
      rowElements.push(
        <tr key={`row_${index}`}
          data-testid={`row_${index}`}>
          <td data-testid={`col_first_${index}`}>
            {row}
          </td>
          <td data-testid={`col_second_${index}`}>
            { product(index, props.number) }
          </td>
        </tr>
      )
    })
    return rowElements
  }

  return (
    <div>
      Products of {props.number} (from 0 to {props.rowCount}):
      <table border={1} bordercolor={'black'}>
        <thead>
          <tr>
            <th>Multiplier</th>
            <th>Product</th>
          </tr>
        </thead>
        <tbody>
          {renderRows()}
        </tbody>
      </table>
    </div>
  )
}

App.propTypes = {
  number: PropTypes.number,
  rowCount: PropTypes.number
}

App.defaultProps = {
  number: 5,
  rowCount: 5
}

export default App