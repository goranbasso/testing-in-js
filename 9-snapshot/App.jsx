import React from 'react'
import PropTypes from "prop-types"

const App = (props) => {
  return (
    <div>
      Snapshot test
      <span>{props.message}</span>
    </div>
  )
}

App.propTypes = {
  message: PropTypes.string
}

App.defaultProps = {
  message: "hei hei hei"
}

export default App