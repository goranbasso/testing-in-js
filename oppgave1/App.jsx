import React from "react";
import PropTypes from "prop-types";

const App = (props) => {
  return <div>Hello {props.name}</div>;
}

App.propTypes = {
  name: PropTypes.string,
}

App.defaultProps = {
  name: "Borld",
}

export default App
