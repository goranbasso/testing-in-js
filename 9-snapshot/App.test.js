import { render } from "@testing-library/react"
import React from "react"
import App from "./App"

/**
 * Dette testsettet er ment å demonstrere hvordan man kan forenkle testingen ved å benytte snapshots.
 *
 * Vi har en enkel applikasjon, som viser noe tekst.
 *
 * For å verifisere at denne vises riktig hver gang, kan vi ta et snapshot av hvordan applikasjonen ser ut,
 * og så benytte dette snapshottet ved å sammenligne det med returnverdien (HTML-koden) som komponentet vårt returnerer.
 * Man kan lagre disse snapshottene til filer, som kan versjonskontrolleres, eller man kan bruke såkalte inline-snapshots,
 * så de står som JSX i selve testkoden.
 * Har valgt å stort sett benytte inline-snapshots her, så det er litt lettere å se hva som foregår.
 *
 * Merk at kjøring av noen av disse testene faktisk vil endre testkoden, ved å legge til dette snapshottet.
 *
 * https://jestjs.io/docs/en/snapshot-testing
 *
 * Kjør følgende kommando i din terminal for å starte:
 * npm run test:watch 9-snapshot
 */

/**
 * Først en test som viser hvordan snapshots ser ut som JSX i testkode. (Det er parameteret til toMatchInlineSnapshot())
 */
it("Rendered app should match stored snapshot", () => {
  const { container } = render(<App />)
  expect(container.firstChild).toMatchInlineSnapshot(`
    <div>
      Snapshot test
      <span>
        hei hei hei
      </span>
    </div>
  `)
})

/**
 * Et eksempel på hvordan man kan generere inline snapshot, til bruk seinere.
 * Ta først å kommenter inn expect(container.firstChild).toMatchInlineSnapshot()-linjen, kjør testen, og se hvordan koden endrer seg.
 */
it("Example on how to generate an inline snapshot", () => {
  const { container } = render(
    <App message={"we're going to generate a new inline snapshot"} />
  )

  // kommenter inn linjen under, og se hvordan testen endrer seg etter en kjøring
  expect(container.firstChild).toMatchInlineSnapshot(`
    <div>
      Snapshot test
      <span>
        we're going to generate a new inline snapshot
      </span>
    </div>
  `)
})

/**
 * Denne testen feiler, fordi det lagrede snapshottet ikke er oppdatert i henhold til koden.
 * Oppdater snapshottet i henhold til koden.
 */
it("Snapshot should be kept up to date", () => {
  const { container } = render(<App message={"the updated snapshot"} />);
  expect(container.firstChild).toMatchInlineSnapshot(`
    <div>
      Snapshot test
      <span>
        the updated snapshot
      </span>
    </div>
  `)
})

/**
 * Vi kan også bruke snapshots til å sjekke at objekter er like.
 * Dette fungerer mer eller mindre på samme måte som for komponenter over.
 */
it("Snapshots can also be used to match object properties", () => {
  const user = {
    createdAt: new Date(328492800000),
    id: 8,
    name: "Steven Gerrard",
  }

  expect(user).toMatchInlineSnapshot(`
    Object {
      "createdAt": 1980-05-30T00:00:00.000Z,
      "id": 8,
      "name": "Steven Gerrard",
    }
  `)
})

/**
 * For enkelte felter er det ikke nødvendigvis så viktig at verdien er helt den samme, men heller at den er av riktig type.
 * Når man bruker property matchere, kan man angi at det er typen, og ikke verdien som er viktig, og at denne blir riktig.
 * https://jestjs.io/docs/en/snapshot-testing#property-matchers
 */
it("Object has the correct types", () => {
  const user = {
    createdAt: new Date(),
    id: Math.floor(Math.random() * 20),
    name: "Roberto Firmino",
  }

  expect(user).toMatchSnapshot({
    createdAt: expect.any(Date),
    id: expect.any(Number)
  })
})

/**
 * Man kan også lagre snapshottene til separate filer, som kan være nyttig for å forenkle test-koden.
 * Skriv en test som sjekker applikasjonen vår opp mot en lagret snapshot-fil.
 * Denne må også genereres for at det skal være mulig å kjøre testen.
 * https://jestjs.io/docs/en/snapshot-testing
 */
it("Rendered app should match snaphot stored in file", () => {
  const { container } = render(<App />)
  expect(container.firstChild).toMatchSnapshot()
})

/**
 * Skriv en lignende test som over, men med egendefinerte verdier inn som props til applikasjonen.
 */
it("Rendered app with custom props should match snapshot stored in file", () => {
  const { container } = render(<App message={"snapshot stored in file"} />)
  expect(container.firstChild).toMatchSnapshot()
})
