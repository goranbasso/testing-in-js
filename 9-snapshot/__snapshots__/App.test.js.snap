// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object has the correct types 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "Roberto Firmino",
}
`;

exports[`Rendered app should match snaphot stored in file 1`] = `
<div>
  Snapshot test
  <span>
    hei hei hei
  </span>
</div>
`;

exports[`Rendered app with custom props should match snapshot stored in file 1`] = `
<div>
  Snapshot test
  <span>
    snapshot stored in file
  </span>
</div>
`;

exports[`Snapshot testing Object has the correct types 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "Roberto Firmino",
}
`;

exports[`Snapshot testing Rendered app should match snaphot stored in file 1`] = `
<div>
  Snapshot test
  <span>
    hei hei hei
  </span>
</div>
`;

exports[`Snapshot testing Rendered app with custom props should match snapshot stored in file 1`] = `
<div>
  Snapshot test
  <span>
    snapshot stored in file
  </span>
</div>
`;
